### PR TITLE
Add -wthread-safety-analysis

### PR DESCRIPTION
**Pull request by danmosemsft**
In https://github.com/dotnet/corefx/issues/29841, and issue was caught using -clang3.9 which referenced -Wthread-safety-analysis. Building on Linux, I cannot seem to repro this issue (before the fix) even applying -clang3.9 and explicitly adding -Wthread-safety-analysis. @janvorli notes that GCC sometimes behaves differently on BSD vs. Linux. So it is possible even without -clang3.9 this flag is already being implicitly applied (it is not being explicitly applied).

Nevertheless it seems reasonable to add the flag explicitly in case it has some value.

@janvorli

### CODE REVIEW

**Comment by janvorli** [comment]
> @janvorli notes that GCC sometimes behaves differently on BSD vs. Linux

You mean clang, not GCC, right? I was commenting on clang.

File: src/Native/Unix/CMakeLists.txt [hunk 1]
```diff
@@ -11,6 +11,7 @@ set(VERSION_FILE_PATH "${CMAKE_BINARY_DIR}/../../version.c")
 
 add_compile_options(-Wno-format-nonliteral)
 add_compile_options(-Wno-disabled-macro-expansion)
+add_compile_options(-Wthread-safety-analysis)
```
Comment (janvorli) [inline src/Native/Unix/CMakeLists.txt]: The option to enable that warning is actually `-Wthread-safety` based on clang doc.
Also, could you please move it out of the `-Wno-xxxx` block here which are warning disabling macros?

Comment (janvorli) [reply src/Native/Unix/CMakeLists.txt]: See https://clang.llvm.org/docs/ThreadSafetyAnalysis.html

**Comment by danmosemsft** [comment]
> You mean clang, not GCC, right? I was commenting on clang.

Yes, thank you.

**Comment by janvorli** [comment]
@danmosemsft interestingly, even with this option added clang doesn't complain on the issue (not even a newer clang 4.0). I was thinking about it and I think it might be due to differences in headers on Linux and FreeBSD. The clang doc on the option mentions some annotating macros that probably need to be used in the system headers on the mutex stuff to make it work. My guess is that FreeBSD headers have them and Linux ones don't.

**Comment by danmosemsft** [comment]
@janvorli yes I tried it and got the same result and came to the same conclusion. It still seems useful as future Linux headers may have such annotations.

**Comment by danmosemsft** [comment]
Would it be worth adding this anywhere in the CoreCLR? Eg., on the PAL?

**Comment by janvorli** [comment]
@danmosemsft I agree. Btw, looking at pthread.h on Linux and FreeBSD:
Linux:
```
/* Lock a mutex.  */
extern int pthread_mutex_lock (pthread_mutex_t *__mutex)
     __THROWNL __nonnull ((1));
```

FreeBSD:
```
#define        __lock_annotate(x)      __attribute__((x))
#define __locks_exclusive(...) \
        __lock_annotate(exclusive_lock_function(__VA_ARGS__))

int             pthread_mutex_lock(pthread_mutex_t *__mutex)
                        __nonnull(1) __locks_exclusive(*__mutex);

```

I guess the __locks_exclusive is either directly understood by clang or gets translated to clang's annotation macros.

**Pull request merged**